src/Tasks/Exec.cs

```diff
@@ -611,18 +624,38 @@ protected internal override void AddCommandLineCommands(CommandLineBuilderExtens
 
                     // If for some crazy reason the path has a & character and a space in it
                     // then get the short path of the temp path, which should not have spaces in it
-                    // and then escape the &
                     if (batchFileForCommandLine.Contains("&") && !batchFileForCommandLine.Contains("^&"))
                     {
                         batchFileForCommandLine = NativeMethodsShared.GetShortFilePath(batchFileForCommandLine);
-                        batchFileForCommandLine = batchFileForCommandLine.Replace("&", "^&");
                     }
+
+                    StringBuilder fileName = StringBuilderCache.Acquire(batchFileForCommandLine.Length);
+
+                    // Escape any characters specified by the CharactersToEscape metadata, or '&'
+                    for (int i = 0; i < batchFileForCommandLine.Length; i++)
+                    {
+                        char c = batchFileForCommandLine[i];
+
+                        if (ShouldEscapeCharacter(c) && (i == 0 || batchFileForCommandLine[i - 1] != '^'))
+                        {
+                            fileName.Append('^');
+                        }
+                        fileName.Append(c);
+                    }
+
+                    batchFileForCommandLine = StringBuilderCache.GetStringAndRelease(fileName);
                 }
 
                 commandLine.AppendFileNameIfNotNull(batchFileForCommandLine);
             }
         }
 
+        private bool ShouldEscapeCharacter(char c)
+        {
+            // Escape '&' to preserve previous functionality
+            return c == '&' || (_charactersToEscape != null && _charactersToEscape.Contains(c));
```

Comment:
I'm thinking about performance here...this function should be fast, but the batch file could be huge, in which case this would be called an immense number of times. It would be nice to factor out things like the null check to make this particular function call even faster.

If you want it even faster, you could make a bit array representing the characters to escape (i.e., bit x is 1 iff (int)c is in _charactersToEscape) and check that instead of this HashSet. On the other hand, I wouldn't be surprised if there's an optimization to HashSet<char> to do just that, so it would be worth checking whether it makes a difference first.

Reply:
Fair point. In theory we could swap the implementation from: "Supply a set of the characters you want escaped" to "Escape all special characters (but if some are already escaped, or you fall into a specific edge case, you're out of luck)".

If we have an array of the characters we know need to be escaped and the user sets "TheMagicFlag", we could iterate through the array and check `if ((c | specialChars[i]) == c)`.

Thoughts? /cc: @ladipro

Reply:
Did some rough testing with stopwatches comparing hashset.contains with a for loop through an array of special characters that checks `if ((c | specialChars[i])==c)`. The loop/array beats out the hashset.contains by several orders of magnitude (in terms of ticks).

Reply:
Here I am on a weekend realizing "wait, why not just do a simple comparison?" I was incepted by the bit comment above 😅

Reply:
Also, (c | specialChars[i] == c) would actually be wrong if, for example, specialChars[i] were 0, it would always return true.